src/main/java/onlinebookstore/service/impl/UserServiceImpl.java

```diff
@@ -0,0 +1,48 @@
+package onlinebookstore.service.impl;
+
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import onlinebookstore.dto.user.UserLoginRequestDto;
+import onlinebookstore.dto.user.UserLoginResponseDto;
+import onlinebookstore.dto.user.UserRegistrationRequestDto;
+import onlinebookstore.dto.user.UserResponseDto;
+import onlinebookstore.exception.RegistrationException;
+import onlinebookstore.mapper.UserMapper;
+import onlinebookstore.model.RoleName;
+import onlinebookstore.model.User;
+import onlinebookstore.repository.RoleRepository;
+import onlinebookstore.repository.UserRepository;
+import onlinebookstore.service.UserService;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+    private final RoleRepository roleRepository;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto registrationRequest)
+            throws RegistrationException {
+        if (userRepository.findByEmail(registrationRequest.email()).isPresent()) {
+            throw new RegistrationException("Unable to complete registration.");
+        }
+        User userToSave = new User();
+        userToSave.setEmail(registrationRequest.email());
+        userToSave.setPassword(passwordEncoder.encode(registrationRequest.password()));
+        userToSave.setFirstName(registrationRequest.firstName());
+        userToSave.setLastName(registrationRequest.lastName());
+        userToSave.setShippingAddress(registrationRequest.shippingAddress());
+        userToSave.setRoles(Set.of(roleRepository.findRoleByName(RoleName.ROLE_USER)));
```

Comment:
For some reason I can't select multiple lines...
So my comments for lines 33-39. Why don't you create a mapper like for the book? ))
You could do like this in your usermapper class: User toUser(UserRegistrationRequestDto user);

And here just do:
User user = userMapper.toUser(registrationRequest);
        user.setRoles(Set.of(roleRepository.findRoleByName(RoleName.ROLE_USER)));
        user.setPassword(encoder.encode(registrationRequest.password()));

So, mapper will do all the dirty work and all you have to do is set the password and roles